### PR TITLE
chore: eliminate go-mpatch dependency

### PR DESCRIPTION
**Pull request by crenshaw-dev**
This test was failing on my Mac due to permissions issues.

I think the modified test is equivalent to the original test. Three methods had been monkey-patched to deterministically generate `serverResources`. I've refactored `generateProjectAllowList` to simply accept the calculated `serverResources` value as a parameter.

Per Leo's suggestion, I modified the non-command functions to return errors instead of exiting. Only the command function should be exiting.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -69,7 +69,6 @@ require (
 	github.com/spf13/cobra v1.3.0
 	github.com/spf13/pflag v1.0.5
 	github.com/stretchr/testify v1.7.0
-	github.com/undefinedlabs/go-mpatch v1.0.6
```
Comment (leoluz) [inline go.mod:72]: 🙏🏻

File: cmd/argocd/commands/admin/project_allowlist.go [hunk 2]
```diff
@@ -63,7 +63,13 @@ func NewProjectAllowListGenCommand() *cobra.Command {
...
 			}
 
-			globalProj := generateProjectAllowList(clientConfig, clusterRoleFileName, projName)
+			config, err := clientConfig.ClientConfig()
+			errors.CheckError(err)
+			disco, err := discovery.NewDiscoveryClientForConfig(config)
+			errors.CheckError(err)
+			serverResources, err := disco.ServerPreferredResources()
+			errors.CheckError(err)
+			globalProj := generateProjectAllowList(serverResources, clusterRoleFileName, projName)
```
Comment (leoluz) [inline cmd/argocd/commands/admin/project_allowlist.go:72]: I don't think the logic to retrieve the resourceList should be added directly in the `NewProjectAllowListGenCommand`. Ideally, commands should be very dry in terms of logic mainly dealing with CLI flags and delegating the execution to an underlying function like it was before. Another possibility to achieve the same goal is defining an interface like:
```go
type ResourceGetter interface {
    GetServerResources(c clientcmd.ClientConfig) ([]*metav1.APIResourceList, error)
}
```
And then have this interface as the first argument of `generateProjectAllowList()`. This way you can easily mock this logic and properly test it.

**Review by leoluz — CHANGES_REQUESTED** [review]
Suggesting a different approach to properly test this.

File: cmd/argocd/commands/admin/project_allowlist.go [hunk 3]
```diff
@@ -78,23 +81,38 @@ func NewProjectAllowListGenCommand() *cobra.Command {
 	return command
 }
 
-func generateProjectAllowList(clientConfig clientcmd.ClientConfig, clusterRoleFileName string, projName string) v1alpha1.AppProject {
+func getResourceList(clientConfig clientcmd.ClientConfig) ([]*metav1.APIResourceList, error) {
+	config, err := clientConfig.ClientConfig()
+	if err != nil {
+		return nil, err
```
Comment (leoluz) [inline cmd/argocd/commands/admin/project_allowlist.go:87]: As we change code that just return `err` I think we should start adding more context to better understand exceptional scenarios in logs. Example:
```go
return nil, fmt.Errorf("error while creating client config: %s", err)
```

File: cmd/argocd/commands/admin/project_allowlist.go [hunk 4]
```diff
@@ -78,23 +81,38 @@ func NewProjectAllowListGenCommand() *cobra.Command {
...
+	}
+	disco, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return nil, err
```
Comment (leoluz) [inline cmd/argocd/commands/admin/project_allowlist.go:91]: please add error context

File: cmd/argocd/commands/admin/project_allowlist.go [hunk 5]
```diff
@@ -78,23 +81,38 @@ func NewProjectAllowListGenCommand() *cobra.Command {
...
+	}
+	serverResources, err := disco.ServerPreferredResources()
+	if err != nil {
+		return nil, err
```
Comment (leoluz) [inline cmd/argocd/commands/admin/project_allowlist.go:95]: please add error context

File: cmd/argocd/commands/admin/project_allowlist.go [hunk 6]
```diff
@@ -78,23 +81,38 @@ func NewProjectAllowListGenCommand() *cobra.Command {
...
+	}
+	return serverResources, nil
+}
+
+func generateProjectAllowList(serverResources []*metav1.APIResourceList, clusterRoleFileName string, projName string) (*v1alpha1.AppProject, error) {
 	yamlBytes, err := ioutil.ReadFile(clusterRoleFileName)
-	errors.CheckError(err)
+	if err != nil {
+		return nil, err
```
Comment (leoluz) [inline cmd/argocd/commands/admin/project_allowlist.go:103]: please add error context

File: cmd/argocd/commands/admin/project_allowlist.go [hunk 7]
```diff
@@ -78,23 +81,38 @@ func NewProjectAllowListGenCommand() *cobra.Command {
...
+	}
 	var obj unstructured.Unstructured
 	err = yaml.Unmarshal(yamlBytes, &obj)
-	errors.CheckError(err)
+	if err != nil {
+		return nil, err
```
Comment (leoluz) [inline cmd/argocd/commands/admin/project_allowlist.go:108]: please add error context

File: cmd/argocd/commands/admin/project_allowlist.go [hunk 8]
```diff
@@ -78,23 +81,38 @@ func NewProjectAllowListGenCommand() *cobra.Command {
...
-	errors.CheckError(err)
-
-	config, err := clientConfig.ClientConfig()
-	errors.CheckError(err)
-	disco, err := discovery.NewDiscoveryClientForConfig(config)
-	errors.CheckError(err)
-	serverResources, err := disco.ServerPreferredResources()
-	errors.CheckError(err)
+	if err != nil {
+		return nil, err
```
Comment (leoluz) [inline cmd/argocd/commands/admin/project_allowlist.go:114]: please add error context

**Review by leoluz — APPROVED** [review]

**Comment by crenshaw-dev** [comment]
Cherry-picked onto 2.3.

**Comment by crenshaw-dev** [comment]
Cherry-picked onto 2.2.

**Pull request merged**